Paper (#399)
mlr-org/mlr3extralearners: paper/paper.md

---

> title: 'mlr3extralearners: A community-driven package for integrating learners into mlr3'

Reviewer: Alternative suggestions: 
1) mlr3extralearners: Expanding the mlr3 Ecosystem with Community-Driven Learner Integration
2) mlr3extralearners: A Unified Interface for Diverse Machine Learning Algorithms in the mlr3 framework

---
title: 'mlr3extralearners: A community-driven package for integrating learners into mlr3'

> tags:

Reviewer: benchmarking?

At its core, the `mlr3extralearners` package provides a standardized interface for machine learning and connects many R packages implementing machine learning algorithms into a unified framework.
The package currently wraps **85 different learning algorithms** from many different R packages, for tasks such as classification, regression, and survival analysis.
This enables users to seamlessly access and utilize these learners directly within their workflows.
It also facilitates large-scale empirical benchmark experiments, leveraging the `mlr3` framework's parallelization and optimization capabilities  [@benchlargescale].
An overview of all `mlr3` learners, including those introduced through `mlr3extralearners`, is available on the [mlr3 website](https://mlr-org.com/learners.html).

Beyond accessibility, `mlr3extralearners` also allows `mlr3` users and package developers to easily add their own learners to the ecosystem.
This **enriches each learner with extensive metadata** about its parameter space, prediction types, and other key attributes.

> Furthermore, `mlr3extralearners` includes robust mechanisms for **quality assurance**, such as regular sanity checks and verification tests that ensure learner parameters are consistent and up-to-date with the latest versions of their underlying R packages.

Reviewer: Maybe be more specific about what are these "sanity checks"?

Furthermore, learners are annotated with respect to their task type (such as classification, regression or survival analysis [@Sonabend2021]) and predict type (such as probabilities or class predictions), which feature types they can handle, and which capabilities they have.
The latter are standardized via a set of standardized properties, which e.g. includes the ability to do feature selection, to assign importance scores to features, or to handle missing values.

## Functional Correctness

One problem that manifests when integrating learning algorithms from different R packages is that their API can change.
The most frequent case is that new parameters are added, which were not present in the version of the package when the learner was integrated.
In `mlr3extralearners`, we regularly check whether the learner implements the expected interface of the upstream function and update the parameter set accordingly.

In addition to this `mlr3`-specific check, `mlr3extralearners` also verifies the correctness of learners by regularly running automatic tests on the learners.

> These tests perform simple sanity checks and also verify that the learner's metadata is correctly annotated, e.g. that a learner that claims to be able to handle missing values actually does so.

Reviewer: maybe be more specific here about the "sanity checks"?

   index: 13
 - name: University of North Carolina Wilmington
   index: 14
 - name: OSPO Now
   index: 15
 - name: Imperial College London
   index: 16
 - name: multi, 8 passage Brûlon, 75012 PARIS, France
   index: 17
date: XXX December 2024
bibliography: paper.bib
---

# Summary

The [`mlr3extralearners`](https://mlr3extralearners.mlr-org.com/) [`R`](https://www.r-project.org/) [@R] package is a community-driven package that integrates external machine learning algorithms into the [`mlr3`](https://mlr3.mlr-org.com/) [@Lang2019] ecosystem.
The `mlr3` ecosystem is a versatile toolbox for machine learning in `R` and is targeted towards both practitioners and researchers [@Bischl2024].
At its core, the `mlr3extralearners` package provides a standardized interface for machine learning and connects many R packages implementing machine learning algorithms into a unified framework.
The package currently wraps **85 different learning algorithms** from many different R packages, for tasks such as classification, regression, and survival analysis.
This enables users to seamlessly access and utilize these learners directly within their workflows.

> It also facilitates large-scale empirical benchmark experiments, leveraging the `mlr3` framework's parallelization and optimization capabilities  [@benchlargescale].

Reviewer 2: If me mention large-scale benchmarks in the abstract, we should add a sentence about tuning.

Reviewer: We mention it in two other spots but I added it also here with a citation, thanks!

Additionally, contributions to `mlr3extralearners` are reviewed by the package maintainers, providing a layer of quality assurance.
This review process ensures that integrated learners work as expected and adhere to the high standards of the `mlr3` ecosystem.

Beyond its utility for users, `mlr3extralearners` also offers significant advantages for developers of machine learning packages.
By integrating a new algorithm into the `mlr3` ecosystem, developers can immediately make their methods accessible to a wider audience.
This integration facilitates seamless tuning [@mlr3tuning] and preprocessing [@mlr3pipelines2021] through the broader `mlr3` framework, enhancing the usability and impact of their work.

# Features

The core functionality of `mlr3extralearners` is to integrate new learners into the `mlr3` ecosystem, allowing users to access a wide array of learning algorithms through a unified syntax and standardized interface.
However, the advantages of `mlr3extralearners` go well beyond simple integration.

> 

Reviewer 2: Why don't we add a few codes examples like [list_mlr3learners()](https://mlr3extralearners.mlr-org.com/reference/list_mlr3learners.html) to show all availble learners? Or show the unified interface with `lrn("classif.gbm")`?

Reviewer: We have the link to the website with all learners, this covers the first. The second we could, let's see what Bernd also thinks when he reads it.

   index: 8
 - name: Department of Statistics and Finance/International Institute of Finance, School of Management, University of Science and Technology of China, Hefei, Anhui, China
   index: 9
 - name: School of Environmental and Forest Sciences, University of Washington, Seattle
   index: 10
 - name: Department of Geography, LMU Munich, Germany
   index: 11
 - name: Statistical Consulting Unit StaBLab, LMU Munich, Germany
   index: 12
 - name: Wake Forest University School of Medicine, Department of Biostatistics and Data Science, Division of Public Health SciencesWinston-Salem, North Carolina
   index: 13
 - name: University of North Carolina Wilmington
   index: 14
 - name: OSPO Now
   index: 15
 - name: multi, 8 passage Brûlon, 75012 PARIS, France
   index: 16
 - name: ServiceTitan, Inc., Glendale, California
   index: 17
date: XXX December 2024
bibliography: paper.bib
---

# Summary

The [`mlr3extralearners`](https://mlr3extralearners.mlr-org.com/) [`R`](https://www.r-project.org/) [@R] package is a community-driven package that integrates external machine learning algorithms into the [`mlr3`](https://mlr3.mlr-org.com/) [@Lang2019] ecosystem.

> The `mlr3` ecosystem is a versatile toolbox for machine learning in `R` and is targeted towards both practitioners and researchers [@Bischl2024].

Reviewer 3: ```suggestion
The `mlr3` ecosystem is a versatile toolbox for machine learning in `R` that is targeted towards both practitioners and researchers [@Bischl2024].
```

> At its core, the `mlr3extralearners` package provides a standardized interface for machine learning and connects many R packages implementing machine learning algorithms into a unified framework.

Reviewer 3: I wouldn't say that this is what "mlr3extralearners" does. The package offering a standardized interface is mlr3. mlr3extralearners is a repository of connectors that integrate existing ML packages into mlr3.

The package currently wraps **85 different learning algorithms** from many different R packages, for tasks such as classification, regression, and survival analysis.
This enables users to seamlessly access and utilize these learners directly within their workflows.
It also facilitates large-scale empirical benchmark experiments, leveraging the `mlr3` framework's parallelization and optimization capabilities [@benchlargescale] as well as enabling efficient hyperparameter tuning [@bischl2023].

> An overview of all `mlr3` learners, including those introduced through `mlr3extralearners`, is available on the [mlr3 website](https://mlr-org.com/learners.html).

Reviewer 3: consider making this a reference or footnote, not a link. When I render the pdf, the link is not visible as such.

Beyond accessibility, `mlr3extralearners` also allows `mlr3` users and package developers to easily add their own learners to the ecosystem.
This **enriches each learner with extensive metadata** about its parameter space, prediction types, and other key attributes.

> Furthermore, `mlr3extralearners` includes robust mechanisms for **quality assurance**, such as regular sanity checks and verification tests that ensure learner parameters are consistent and up-to-date with the latest versions of their underlying R packages.

Reviewer 3: ```suggestion
Furthermore, `mlr3extralearners` includes robust mechanisms for **quality assurance**, such as regular automated sanity checks and verification tests that ensure learner parameters are consistent and up-to-date with the latest versions of their underlying R packages.
```

> In order to allow the integration of learners that are not available on `CRAN`, the package is hosted on the [`mlr` R-universe](https://mlr-org.r-universe.dev/).

Reviewer 3: same as above with the link: maybe this should be a reference or footnote.

Reviewer 3: Besides, our reason for not being on CRAN are somewhat different, aren't they? Putting it in the best possible light, we could say that we want to be able to react quickly to changes in the many package dependencies that we have, which does not fit to the somewhat heavyweight release model of CRAN.

By providing a standardized interface and comprehensive metadata for each learner, mlr3extralearners enhances the FAIRness (findability, accessibility, interoperability, and reusability) of machine learning algorithms within the R ecosystem [@wilkinson2016fair]).

# Statement of Need

> Machine learning often requires practitioners to navigate a diverse array of modeling problems, each with unique demands such as performance, interpretability, or compatibility with specific data types and tasks.

Reviewer 3: ```suggestion
Machine learning often requires practitioners to navigate a diverse array of modeling problems, each with unique demands such as predictive performance, inference latency and throughput, interpretability, or compatibility with specific data types and tasks.
```

Additionally, contributions to `mlr3extralearners` are reviewed by the package maintainers, providing a layer of quality assurance.
This review process ensures that integrated learners work as expected and adhere to the high standards of the `mlr3` ecosystem.

Beyond its utility for users, `mlr3extralearners` also offers significant advantages for developers of machine learning packages.
By integrating a new algorithm into the `mlr3` ecosystem, developers can immediately make their methods accessible to a wider audience.
This integration facilitates seamless tuning [@mlr3tuning] and preprocessing [@mlr3pipelines2021] through the broader `mlr3` framework, enhancing the usability and impact of their work.

# Features

The core functionality of `mlr3extralearners` is to integrate new learners into the `mlr3` ecosystem, allowing users to access a wide array of learning algorithms through a unified syntax and standardized interface.
However, the advantages of `mlr3extralearners` go well beyond simple integration.

## Metadata

One core feature of the `mlr3` ecosystem is that it annotates learners with extensive metadata.

> - **Parameter Management**: The parameter spaces of learners are defined using parameter sets from the [`paradox` package](https://paradox.mlr-org.com/) [@paradox].

Reviewer 3: ```suggestion
- **Hyperparameter Management**: The hyperparameter spaces of learners are defined using `ParamSet` objects from the [`paradox` package](https://paradox.mlr-org.com/) [@paradox].
```

> Each parameter is explicitly typed, with annotations for valid ranges and allowable values.

Reviewer 3: ```suggestion
Each hyperparameter is explicitly typed, with annotations for valid ranges and allowable values.
```

> This ensures valid configurations and simplifies tasks like parameter tuning.

Reviewer 3: ```suggestion
This ensures valid configurations and simplifies tasks like hyperparameter tuning.
```

> - **Task and Prediction Types**: Learners are categorized with respect to their task type (e.g. as classification, regression or survival analysis [@Sonabend2021]) and prediction types (e.g. probabilities or class predictions).

Reviewer 3: ```suggestion
- **Task and Prediction Types**: Learners are categorized with respect to their task type (e.g. as classification, regression or survival analysis [@Sonabend2021]) and prediction types (e.g. probabilities or response predictions).
```

This allows users to easily identify suitable learners for their specific modeling tasks.

> - **Standardized Properties**: Learners are annotated with detailed attributes, including the types of features they can process and their support for functionalities such as feature selection such as feature selection, importance scoring, handling missing values, or monitoring performance on a separate validation set during training among others.

Reviewer 3: ```suggestion
- **Standardized Properties**: Learners are annotated with detailed attributes, including the types of features they can process and their support for functionalities such as feature selection, importance scoring, handling missing values, or monitoring performance on a separate validation set during training among others.
```

This allows users to have a clear understanding of a learner's capabilities and limitations and assess if it aligns with the specific requirements of their workflows, reducing trial-and-error and streamlining the modeling process.

## Functional Correctness

Integrating learners from diverse R packages poses challenges, particularly due to changes in upstream APIs. `mlr3extralearners` addresses these issues through rigorous checks:

- **Interface Consistency**: The package regularly verifies that each learner adheres to the expected interface of its upstream function.

> When new parameters are introduced or existing ones change, the tests fail until the parameter sets are updated accordingly.

Reviewer 3: Do our automatic checks actually notice when upstream introduces new parameters? Or only when they change / are removed?

- **Automated Testing**: To ensure correctness, `mlr3extralearners` performs regular automated tests on all learners.
These tests include sanity checks that, e.g., verify that the learners produce sensible predictions for simple tasks.
Furthermore, the tests also validate the learners' metadata annotations, such as whether a learner can actually handle missing values or is able to produce importance scores.

## Simplified Integration of New Learners

To streamline the addition of new learners, `mlr3extralearners` provides robust support tools:

- **Code Templates**: Predefined templates are available for generating the necessary code for both the learner implementation and associated test files.

> These templates are generated through an `R` function that uses learner metadata to prefill as much information as possible, leaving only minimal input required from the contributor.

Reviewer 3: This part is a bit confusing with regard to what the template does. Is the template the thing that is used to generate code, or is the template itself generated?

> - **Guides and Resources**: The package website contains an contains an [extensive tutorial](https://mlr3extralearners.mlr-org.com/articles/extending.html), as well as a curated list of [common issues](https://mlr3extralearners.mlr-org.com/articles/common_issues.html) encountered during learner integration, making the process accessible for contributors of all experience levels.

Reviewer 3: ```suggestion
- **Guides and Resources**: The package website contains an [extensive tutorial](https://mlr3extralearners.mlr-org.com/articles/extending.html), as well as a curated list of [common issues](https://mlr3extralearners.mlr-org.com/articles/common_issues.html) encountered during learner integration, making the process accessible for contributors of all experience levels.
```

   index: 4
 - name: Leibniz Institute for Prevention Research and Epidemiology (BIPS), Bremen, Germany
   index: 5
 - name: Faculty of Mathematics and Computer Science, University of Bremen, Germany
   index: 6
 - name: Certilytics, Inc., 9200 Shelbyville Rd, Louisville, KY, 40222, USA
   index: 7
 - name: Friedrich-Alexander-Universität Erlangen-Nürnberg (FAU), Erlangen, Germany
   index: 8
 - name: Department of Statistics and Finance/International Institute of Finance, School of Management, University of Science and Technology of China, Hefei, Anhui, China
   index: 9
 - name: School of Environmental and Forest Sciences, University of Washington, Seattle
   index: 10
 - name: Department of Geography, LMU Munich, Germany
   index: 11
 - name: Statistical Consulting Unit StaBLab, LMU Munich, Germany
   index: 12
 - name: Wake Forest University School of Medicine, Department of Biostatistics and Data Science, Division of Public Health SciencesWinston-Salem, North Carolina
   index: 13
 - name: University of North Carolina Wilmington
   index: 14
 - name: OSPO Now
   index: 15
 - name: Imperial College London

>    index: 16

Reviewer 4: I'm no longer affiliated with Imperial so probably best to remove, thanks

   index: 11
 - name: Statistical Consulting Unit StaBLab, LMU Munich, Germany
   index: 12
 - name: Wake Forest University School of Medicine, Department of Biostatistics and Data Science, Division of Public Health Sciences Winston-Salem, North Carolina
   index: 13
 - name: University of North Carolina Wilmington
   index: 14
 - name: OSPO Now
   index: 15
 - name: multi, 8 passage Brûlon, 75012 PARIS, France
   index: 16
 - name: ServiceTitan, Inc., Glendale, California
   index: 17
 - name: devXY GmbH
   index: 18
date: 24 March 2025
bibliography: paper.bib
---

# Summary

The `mlr3` ecosystem is a versatile toolbox for machine learning (ML) in `R` [@R] that is targeted towards both practitioners and researchers [@Bischl2024].
The core `mlr3` package [@Lang2019] defines the standardized interface for ML, but its goal is not to implement algorithms.
This is, e.g., done by the `mlr3learners` extension [@mlr3learners] that connects 21 stable learning algorithms from various `R` packages to the `mlr3` ecosystem that serve as a good starting point for many ML tasks.
In addition, `mlr3extralearners` is a *community-driven* package that integrates many more methods.

> The package currently wraps **85 different ML algorithms** from many different `R` packages, for tasks such as classification, regression, and survival analysis.

Author: did you count only those from mlr3extralearners or all learners?
